### ISSUE TITLE
This library is not usable unless/until CDS plugins properly support it

### ISSUE THREAD

**Issue by fberasategui** [microsoft/PowerPlatform-CdsServiceClient#11]
All of my CDS-based projects (i.e all projects of my company, since we exclusively provide CDS-based solutions, which I love) look like this:

```
 - MyProject.Model
 - MyProject.BusinessLogic
 - MyProject.Plugins
 - MyProject.Api
 - Etc...
```

The `Model` project will contain early-bound entities which derive from `Microsoft.Xrm.Sdk.Entity`.  Then, the `BusinessLogic` layer contains all business logic for the entire project, which uses the entities defined above. Then the rest of the solution will consume these 2.

But now I have 2 versions of `Microsoft.Xrm.Sdk.Entity`, from 2 different assemblies, that I need to derive from depending on whether I'll be using them for a plugin or for an external component such as a Web API, Azure WebJob or Function, etc. This effectively means that I now have 2 different, incompatible set of entities, which completely destroy all possibilities of reusing anything or having a common base layer.

In an ideal scenario (apart from running the entire CDS platform on .NET Core, which seems far fetched), this library should target .NET Standard 2.0 and be compatible with both .NET Framework 4.x assemblies (which can then be used for plugins) **AND** .NET Core 3.0, for any usage anywhere outside the CDS platform, such as Azure Functions and so on.

**Comment by jordimontana82** [comment]
Hi Fede,

Agree, but I think @MattB-msft mentioned that there were some technical limitations by which they had to go for a multi target package instead of targeting .net standard...  given said that...

One workaround is to put your model / business logic projects  into one multi target nuget package each... in that way you could consume the same package both from client applications (api, aspnet core, azure functions in .net core), AND server applications would be linked to the "old" SDK Tooling in full .net framework. For example, rather than using shared projects, we produce many nuget packages internally with both .dll's and source code, they're the same package, but one is aimed at external applications (.dlls) and the source package aimed at plugins / codeactivities etc.

Now, this is because even though they're different assemblies they still share pretty much the same namespaces...  but things will start to break as soon as the new CdsServiceClient repo starts to divert from the existing XrmTooling....

**Comment by fberasategui** [comment]
@jordimontana82 Thanks for the suggestion.

But that would mean I have to split up all of my company-wide nugets into 14 different packages.

Also: business logic does not belong into a nuget package. Business logic needs to be easy to modify and extend, and the write code -> CI -> update nuget package loop is terribly unsuitable for code that changes regularily

**Comment by MattB-msft** [comment]
.Net core support for plugins is blocked by this client at this time,  as we are not in a position to support that server side and decided to head off the bad experience before customers and developers ran into it. 

Adding to that is that we do intended to diverge Namespaces over the next year or so, similar to what you see in this package now,  where the base package only has CDS core Namespaces.   This will father change over time as the base namespace Microsoft.Xrm begins to change. 

But to be clear,  these nuget packages are intended for Client development only,  not for server development.

**Comment by fberasategui** [comment]
@MattB-msft so what's the recommended approach to have a base layer that shares model and business logic for a CDS project that requires both plugin-based and external (Azure, etc) functionality?

**Comment by jordimontana82** [comment]
@fberasategui 14 packages? why? I mean a single package can multi target both full .net framework and .net core. Even if you expose a project as 2 different packages (i.e. dll's and one being source code) it's still one codebase. 

And as per the push -> CI -> terrible experience , depends on how long the build takes, but you could still push your packages to your local file system as a separate nuget package source and that is FAST for local dev, while also having the normal nuget server package source for CI builds :) 

My main kind of problem  is that even with multi target the namespaces will start to divert at some stage so.... but not because it doesn't target .net standard, but because is still an alpha package and likely to change (which is normal)

**Comment by fberasategui** [comment]
> 14 packages? why? I mean a single package can multi target both full .net framework and .net core. Even if you expose a project as 2 different packages (i.e. dll's and one being source code) it's still one codebase.

Because I have at least 7 packages on top of `Microsoft.CrmSdk.CoreAssemblies` at the moment. If I need to split these up so that I have 2 versions of each (one for .NET and one for .NET Core), I end up with effectively 14 different packages.

> And as per the push -> CI -> terrible experience , depends on how long the build takes,

Yeah, sorry, No. I'm not going to put my business logic into nugets. That's simply wrong and unacceptable. Some of my devs are NOT authorized to even touch the csproj file. There's no way I'm going to force them to deal with nuget stuff only to create a plugin that modifies some value from some entity.

if suppose I'll need to start dealing with compilation constants, conditional compilation, and multi targetting, if there's no better way to deal with this.

**Comment by jordimontana82** [comment]
> Because I have at least 7 packages on top of `Microsoft.CrmSdk.CoreAssemblies` at the moment. If I need to split these up so that I have 2 versions of each (one for .NET and one for .NET Core), I end up with effectively 14 different packages.

So... Multitarget does not need 2 packages, the one package can multi target both frameworks, [here's an example](https://www.cafe-encounter.net/p2312/multi-targetting-net-framework-and-net-core-in-a-single-project):

> Yeah, sorry, No. I'm not going to put my business logic into nugets. That's simply wrong and unacceptable

Don't agree with this bit at all, from my own experience, there are many cases where business logic can be reused by at least 2 consumers and so it makes sense to put it in a nuget package..., if the change is small enough probably not worth it, fair enough, but from there to saying is completely unacceptable under any circumstance....  there's a big difference.

So yeah, multi target for the time being OR wait for a more stable release :)

**Comment by fberasategui** [comment]
> So... Multitarget does not need 2 packages, the one package can multi target both frameworks, here's an example:

Yeah, it deals with conditional compilation, conditional references, polyfills, and a bunch of stuff that I shouldn't really have to care about. But from @MattB-msft's comment I'm guessing it's the only way forward. 

Which is kind of lame anyways because plugin development is going to be stuck in an obsolete version of the .NET platform and the C# language. Many C# 8.0 and above language features require .NET Core for some reason (such as default implementations)

> Don't agree with this bit at all,

Using package references between projects in the same solution is stupid and terribly wasteful. There's no denying that. You're putting an unnecessary level of indirection between two software components that should have a direct reference. Reminds me of the microservice fanatics who put an HTTP layer between regular classes for no reason, just because it's the "modern" way.

**Comment by jordimontana82** [comment]
> Using package references between projects in the same solution is stupid and terribly wasteful. There's no denying that

Excuse me but...  can you please tell me where and when I said "same solution" ?  I said 2 consumers... which is perfectly fine for reasonable non trivial logic that spans many repos, for example ...

**Comment by fberasategui** [comment]
@jordimontana82 the entire discussion is based on my first post, where I specify the usual solution structure I use for all my CDS based software projects.

I fully agree with you that reusable logic should be put into Nugets, IF and only if it needs to span multiple solutions/repos, but that's not the case I'm talking about in this thread.

**Comment by MattB-msft** [comment]
@fberasategui,  
thanks for your feedback. 
a few comments. 

> Which is kind of lame anyways because plugin development is going to be stuck in an obsolete version of the .NET platform and the C# language. Many C# 8.0 and above language features require .NET Core for some reason (such as default implementations)

If you will allow a different interpretation; 
We are running on a .net framework that is bedrock solid, we know how it preforms, how it breaks and where its limits are.  its a low enough version that as we upgrade customers from onpremise, we do not require them to do mass recompiles of code to install plugins or talk to us.  its also fully supported on .4.7.2 and .4.8. Thus seeking maximum compatibility and reliability at the detriment of being on the leading edge. (.net 4.6.2 in active support and will remain there for a number of years yet.)   

Given that, We are using this client and its modifications to validate our .net core support and feature set for our strongly types SDK/API infrastructure as we move around or Infrastructure to support updating various bits of it to .net core and like languages. 

We do offer direct access to our WebAPI for folks that want to go down to the wire protocol and develop there own clients. 

> @MattB-msft so what's the recommended approach to have a base layer that shares model and business logic for a CDS project that requires both plugin-based and external (Azure, etc) functionality?

As this client is still in Alpha we have not formalized that guidance, that said, we are doing this internally and are working out the kinks/bugs along the way.

**Comment by fberasategui** [comment]
@MattB-msft thanks for your response.

I fully understand all the back-compat arguments.

> Given that, We are using this client and its modifications to validate our .net core support and feature set for our strongly types SDK/API infrastructure as we move around or Infrastructure to support updating various bits of it to .net core and like languages.

This "reboot" looks like a huge opportunity to provide a richer API and better abstractions as discussed [here](https://github.com/microsoft/PowerPlatform-CdsServiceClient/issues/15). It would be awesome to hear your opinion on that.

> We do offer direct access to our WebAPI for folks that want to go down to the wire protocol and develop there own clients.

I took your suggestion to not try to reimplement `IOrganizationService` targetting the Web API, so I was hoping that this new SDK was going to be the chance for me to move all my infra to .NET Core. For instance, I'm still using Azure Functions V1, and there are lots of improvements that I'm missing from V2, but I can't use it as long as I depend on `Microsoft.CrmSdk.CoreAssemblies`.

But now I have to weight out the benefits of moving my functions to Core versus the amount of work and permanent maintenance overhead involved in having to target 2 different versions of the SDK, instead of just upgrading everything and calling it a day.

**Comment by daryllabar** [comment]
I'm going late to this conversation, but if you're using Visual Studio, then would a Shared Project work for allowing for code reuse between different projects?  Or are Shared Projects not supported in .net core?

**Comment by jordimontana82** [comment]
@daryllabar They are, in full VS Ide though, not in VS Code unless there is a extension for that (probably?)

**Comment by jordimontana82** [comment]
Afaik they are @daryllabar , with full IDE >= 2017 at least.  Not sure about VSCode, which is what I use now (for .net core stuff)

**Comment by MattB-msft** [comment]
Closing this up for now.. the API and SDK teams are looking at this for future planning, thanks for the great feedback and discussion here.
